tests/middleware/fieldPermissions.test.js

```diff
@@ -0,0 +1,267 @@
+  supertest(server)
+  .put(`/test/events/${eventId}`)
+  .send({
+    f1: 'a',
+    f2: 'a',
+    f3: 'a',
+    f4: 'a',
+    f5: 'a',
+    f6: 'a'
+  })
+  .set('Authorization', 'Bearer ' + adminToken)
+  .expect(401)
+  .then(res => {
+    // res.body should be a message including which fields are unauthorized
+    t.ok(
+      res.body.message.includes('f1'),
+      'message should list fields which are unauthorized')
+    t.notOk(
+      ['f2', 'f3', 'f4', 'f5', 'f6'].some(field => res.body.message.includes(field)),
+      'message should not list fields which are authorized')
+    t.end()
+  })
+  .catch(err => t.end(err))
+})
+
+tape('fieldPermissions PUT editing various fields, as basic user and not owner', t => {
+  supertest(server)
+  .put(`/test/events/${eventId}`)
+  .send({
+    f1: 'a',
+    f2: 'a',
+    f3: 'a',
+    f4: 'a',
+    f5: 'a',
+    f6: 'a'
+  })
+  .set('Authorization', 'Bearer ' + basicToken)
+  .expect(401)
+  .then(res => {
+    // res.body should be a message including which fields are unauthorized
+    t.ok(
+      ['f1', 'f2', 'f4', 'f5'].every(field => res.body.message.includes(field)),
+      'message should list fields which are unauthorized')
+    t.notOk(
+      ['f3', 'f6'].some(field => res.body.message.includes(field)),
+      'message should not list fields which are authorized')
+    t.end()
+  })
+  .catch(err => t.end(err))
+})
+
+// 4) tests for individual functions
+
+tape('test getUnauthorizedFields with no fields', t => {
+  const user = {}
+  const fieldPermissions = {}
+  const fieldsToChange = []
+
+  const unauthorizedFields =
+    getUnauthorizedFields(fieldPermissions)(fieldsToChange)(user)
+
+  t.equal(unauthorizedFields.length, 0, 'no unauthorized fields')
+  t.end()
+})
+
+tape('test getUnauthorizedFields with varied fields', t => {
+  const user = {
+    role: roles.BASIC,
+    isResourceOwner: true
+  }
+  const user2 = {
+    role: roles.ADMIN
+  }
+
+  const fieldPermissions = {
+    'f1': [ roles.SUPER ],
+    'f2': [ roles.ADMIN ],
+    'f3': [ roles.BASIC ],
+    'f4': [ roles.SUPER, roles.OWNER ],
+    'f5': [ roles.ADMIN, roles.OWNER ],
+    'f6': [ roles.BASIC, roles.OWNER ]
+  }
+  const fieldsToChange = ['f1', 'f2', 'f3', 'f4', 'f5', 'f6']
+
+  const getFields =
+    getUnauthorizedFields(fieldPermissions)(fieldsToChange)
+
+  t.deepEqual(
+    getFields(user),
+    ['f1', 'f2'],
+    'should return correct unauthorized fields for basic user who is resource owner'
+  )
+
+  t.deepEqual(
+    getFields(user2),
+    ['f1', 'f4'],
+    'should return correct unauthorized fields for admin user'
+  )
+  t.end()
+})
+
+tape('fieldPermissions initialization with bad implementation', t => {
+  try {
+    fieldPermissions({
+      id: [ roles.SUPER, roles.ADMIN ]
+    })
+    t.end('should throw bad implementation')
+  } catch (err) {
+    t.end()
+  }
+})
+
+tape('fieldPermissions initialization', t => {
+  try {
+    fieldPermissions({
```

Comment:
test with assertion can be confusing. `t.doesNotThrow`?

Reply:
I didn't know about `t.throws` and `t.doesNotThrow`, that's what I was after.

Reply:
changed